### ISSUE TITLE
Allow moving default master with avar2

### ISSUE THREAD

**Issue by behdad** [harfbuzz/boring-expansion-spec#89]

**Comment by behdad** [comment]
cc @nedley I would appreciate your feedback.
cc @justvanrossum

**Comment by behdad** [comment]
cc @frankrolf

**Comment by Lorp** [comment]
@behdad while there may be strong use cases for "moving the default" as you describe using avar2, it seems to me that the frequently requested use case you describe is much better handled in `fvar`. For example, we can take one of the 16 _InstanceRecord_ flags to declare that this instance should be used as the default instance.

A separate issue is that perhaps we need a new name for this kind of instance, such as "representative", because "default" sounds too much like "fallback", and there’s a lot of documentation defining default instance as what is used when variations are unavailable.

**Comment by behdad** [comment]
> @behdad while there may be strong use cases for "moving the default" as you describe using avar2, it seems to me that the frequently requested use case you describe is much better handled in `fvar`. For example, we can take one of the 16 _InstanceRecord_ flags to declare that this instance should be used as the default instance.

That's an interesting approach.

**Comment by justvanrossum** [comment]
> use case you describe is much better handled in fvar

I'm not sure that's true. Each fvar axis has a default value, and that is a fine user-facing concept. Assigning a "default location" via the instances fundamentally goes against that concept.

**Comment by justvanrossum** [comment]
> A separate issue is that perhaps we need a new name for this kind of instance, such as "representative", because "default" sounds too much like "fallback", and there’s a lot of documentation defining default instance as what is used when variations are unavailable.

Then we'd have two names for two related but subtly different concepts. I think that's a very bad idea.

**Comment by Lorp** [comment]
> Then we'd have two names for two related but subtly different concepts. I think that's a very bad idea.

Hmm, but @Behdad’s proposal already undermines one well-established meaning for Default, being the instance you get when variations are unavailable. Is that meaning already redundant?

> Each fvar axis has a default value, and that is a fine user-facing concept.

What are the disadvantages of treating the Representative instance as the Default in apps?

**Comment by justvanrossum** [comment]
> one well-established meaning for Default, being the instance you get when variations are unavailable

I don't think this meaning is very relevant anymore for end users.

The requested feature fundamentally cannot work in situations when variations are unavailable, so I am not sure how relevant the observation is.

We don't want to change anything from the user's perspective _in a vf-savvy environment_, we just want to make it possible to build fonts that don't need a master at the default to actually not have a master at the default. The user-facing parts of the font shouldn't change: you'd give a user an updated replacement font and all they would notice is that the font file is smaller.

**Comment by Lorp** [comment]
> We don't want to change anything from the user's perspective in a vf-savvy environment, we just want to make it possible to build fonts that don't need a master at the default to actually not have a master at the default. The user-facing parts of the font shouldn't change: you'd give a user an updated replacement font and all they would notice is that the font file is smaller.

I’m with you all the way on this wish! I’m simply proposing the same result with less maths.

**Comment by nedley** [comment]
If my understanding is correct, the request here is to allow avar2 (or something else, perhaps an `fvar` flag) to enable variation processing for what would otherwise be the default instance; the claimed use case for this is to allow a variable font to be regenerated with fewer deltas. Since there’s no way of accomplishing this without producing a different default instance for clients that don't support variations, my question is: why not go ahead change the default instance from wght=400 to 100?

**Comment by frankrolf** [comment]
This is based on a suggestion I made at the recent ATypI panel on Variable Fonts.

My main concern – in most cases, a corner of the VF designspace is not a suitable representative of the family as a whole. Usually, the Regular (or equivalent) is at the core of the design, which is the style that will also get the most use.

What I was hoping for is a flag (for VF-savvy systems) that indicates “this style is my design representative”.
As such, my suggestion was based on aesthetic considerations, not technical reasons.

**Comment by behdad** [comment]
cc @skef

**Comment by skef** [comment]
Thanks @behdad 

Adobe also favors using an `fvar` InstanceRecord bit for this purpose. Here's our write-up on this that we'll probably start circulating elsewhere: [vfdefault.pdf](https://github.com/harfbuzz/boring-expansion-spec/files/12164416/vfdefault.pdf)

Note that we don't think this should necessarily preclude other options, such as avar remapping, for cases where that makes sense. We *do* think that there should be a simpler option for font engineers that don't want to tread into those deep waters.

In researching this the main objection to using an fvar instance seems to be that some people object to that mechanism generally, in favor of this or that kind of entry in `STAT`. My tentative response to that concern would be:

1. `fvar` instances seem to be the most salient means of picking out a particular point in design space
2. If some people would prefer that particular InstanceRecords only be used for particular cases (such as setting what we've called the "interface default", or allowing access to an instance via a string, or presenting an instance in a UI), we should figure out how to separate out those cases with additional flag bits.

**Comment by behdad** [comment]
> Adobe also favors using an `fvar` InstanceRecord bit for this purpose. Here's our write-up on this that we'll probably start circulating elsewhere: [vfdefault.pdf](https://github.com/harfbuzz/boring-expansion-spec/files/12164416/vfdefault.pdf)
> 
> Note that we don't think this should necessarily preclude other options, such as avar remapping, for cases where that makes sense. We _do_ think that there should be a simpler option for font engineers that don't want to tread into those deep waters.

Thanks @skef. I agree that `fvar` flag makes most sense. I still want to relax some conditions on avar2 if feasible.

**Comment by Lorp** [comment]
Adding link to TypeDrawers discussion [*Are customers buying or using variable fonts?*](https://typedrawers.com/discussion/comment/60852/#Comment_60852) where the idea of "representative style" and "representative instance" is raised.

Some thought is needed for recommended behaviour in CSS. Should the Default instance or the Representative instance be used when no axis locations are specified? And what if some axis locations are specified but not others? Should unspecified axes default to Default or should they default to Representative?

**Comment by behdad** [comment]
> Some thought is needed for recommended behaviour in CSS.

HTML text always has currently-set weight/width/slant/opsz. So it doesn't care about the default instance as long as those axes exist.

This whole issue is mostly moot on most systems.

**Comment by skef** [comment]
Tentatively:

1. "Should the Default instance or the Representative instance be used when no axis locations are specified?" The representative instance
2. "And what if some axis locations are specified but not others? Should unspecified axes default to Default or should they default to Representative?" Again to representative, I think. Adding some data should be continuous with adding no data.

**Comment by Lorp** [comment]
@behdad Good point, browsers already set wght=400 (or 700 if in H1, etc.), wdth=100, opsz=<currentPxSize>, which should override whatever is in Representative Instance. So it’s only other axes that this question applies to. See https://codepen.io/lorp/pen/MWzqxej

Apps following standard browser behaviour — by applying sensible defaults for registered axes — would fix a lot of the practical problems here, in which case the Representative Instance flag would only be necessary for unregistered axes. Google’s separately maintained axis registry could conceivably also become respected by apps in terms of default behaviour.

**Comment by justvanrossum** [comment]
FWIW, I'd like to rephrase why I don't like the "representative instance" idea: we already have a perfectly good mechanism _from the consumer's perspective_ to specify the default, which is the default axis values in fvar. With a "representative instance" flag, we add a new mechanism on top of the old mechanism for the sole reason that the _implementation_ of the old mechanism is incapable of a desired feature. So instead of fixing the underlying problem, the "representative instance" adds another layer of complexity _for the consumer_. 

It is adding a new API just because it is difficult to fix the _implementation_ of the (identical) existing API.

To me this feels very much like the "name IDs 1 and 2 have implementation limits that we don't want to fix, so let's some more name IDs" way of thinking.

I will admit that there is a possibility that the initial cost of fixing the existing implementations is higher that adding this new API, but having a more complex surface area for consumers is something that will stay forever, while implementation fixes are a one-time thing.

**Comment by skef** [comment]
@justvanrossum Can you say more about how this addition would wind up being consumer-visible? It seems to me that the goal is to allow a variable font that doesn't have its format default (all axes 0) at the desired interface default/starting point to act the same way as a font that does, and the flag proposal would meet that goal. I'm not aware of how a user would even be aware of the difference short of digging into the font file. 

As for the meaning of "default" in the axis description in fvar, that seems more like misleading documentation to me than anything. The value put there is always interpreted as corresponding to axis point 0 and should probably be described as such. (The description, at least in the Microsoft docs, makes it sound a bit like you're picking the default *location* of the axis by setting the value, when in fact you're picking the numeric value to assign the location you've already picked.)

**Comment by justvanrossum** [comment]
> (Or maybe by "consumer" you mean system-level font software?)

Yes.

> As for the meaning of "default" in the axis description in fvar, that seems more like misleading documentation to me than anything. The value put there is always interpreted as corresponding to axis point 0 and should probably be described as such. (The description, at least in the Microsoft docs, makes it sound a bit like you're picking the default location of the axis by setting the value, when in fact you're picking the numeric value to assign the location you've already picked.)

I'm not sure I follow.

Here we should be talking about end users. What does "axis point 0" mean to an end user? fvar axes are the translations from _end user visible_ values to the internals (normalized space), so the "default axis value" is a meaningful bit of information: "if you don't touch this dial, this is where you'll be" (with CSS overruling that idea for wght and wdth), and I think VF UIs should (and generally do) respect that information, by initializing all axis dials to the default value.

**Comment by justvanrossum** [comment]
Btw. another way to achieve "user-level default should not imply font-implementation-varition-model-default" is by lifting the *avar2* restriction that 0 must map to 0 in `axisValueMaps`.

**Comment by skef** [comment]
> so the "default axis value" is a meaningful bit of information: "if you don't touch this dial, this is where you'll be" (with CSS overruling that idea for wght and wdth), and I think VF UIs should (and generally do) respect that information, by initializing all axis dials to the default value.

Sure, but that leaves open the question of what that default is for a given font. If the user has no idea that there is a format-internal 0 value for each axis then there's no particular significance to having the default correspond to 0 versus something else.

I partially sympathize with the "not repairing what's broken" argument you're making, but I think I disagree about what is broken and what the current specification is really saying. Arguably *nothing* is broken. The current spec puts special significance on 0 axis values. Some people want to build fonts differently because they have corner masters and interpolating an extra master to effect the desired default instance makes the file bigger. The current spec says is, in effect, "Don't do that. Interpolate an extra master for that location and accept the added size." *That's* the purity argument.

Some people want to build variable fonts without that restriction, because file size is relevant. So what we're exploring is how to change things to allow that. I don't personally see "specification purity" in the avar 2 answer to the question. I also think that it's a complicated answer to a simple problem, and I worry that if some font engineers are forced to use it for this purpose they'll no longer fully understand the fonts they're creating.

**Comment by justvanrossum** [comment]
So, is your main point that you don't want to have to use avar2 _just_ for not having a master at the location implied by the fvar default axis values?

**Comment by skef** [comment]
Yes. Moving 0 with avar 2 is "graduate level" font engineering. Some folks just want to build their 2 or 4 master fonts.

**Comment by skef** [comment]
(Which, it's worth noting, actually works just fine now *except* for this problem of initial axis values. All the other stuff remains entirely workable with only "extreme" masters.)

**Comment by justvanrossum** [comment]
The solution (with modified avar2 or even modified avar1) would be trivial to add to fontmake, so I don't believe the font engineering difficulty argument.

**Comment by justvanrossum** [comment]
> The solution (with modified avar2 or even modified avar1) would be trivial to add to fontmake, so I don't believe the font engineering difficulty argument.

I meant to concretely describe this: if a designspace file does not specify a source at the default location for all sources, fontmake will complain. It could instead pick another source as the internal default source, and apply avar mapping to fix.

**Comment by skef** [comment]
And you're saying that the font engineer will still have reasonable intuition for how the user space maps to the internal space space when debugging issues, etc?

**Comment by skef** [comment]
Because if the idea is "the compiler just takes care of things and they work or you go back and fiddle with the sources" I don't think we're there yet.

**Comment by justvanrossum** [comment]
I would expect that preparing sources for a VF that doesn't have a source for the default fvar location to be not harder than to specify a regular old avar1-style mapping in a designspace file.

**Comment by skef** [comment]
> I would expect that preparing sources for a VF that doesn't have a source for the default fvar location to be not harder than to specify a regular old avar1-style mapping in a designspace file.

I'm not sure how this answers my question.

**Comment by anthrotype** [comment]
> [fontmake] could instead pick another source as the internal default source, and apply avar mapping to fix.

the problem with that is how should the compiler pick the other default source among those available, given that (quoting @behdad https://github.com/googlefonts/fontc/pull/345#issuecomment-1625671191)

> You'd get different interpolation results based on which master you pick as base.

Besides the default source is by definition "full" (not sparse) so anything _not_ in it which may be present elsewhere is as if not there at all.

**Comment by justvanrossum** [comment]
(Ok, an implicit base source is clearly problematic, forget I suggested it.)

I think we have three options for "moving the default instance":
- Add a representative instance flag
- Modify the avar2 mechanism
- Modify the **avar1** mechanism, by loosening this part of the spec:
  `If the segment map for a given axis has any value maps, then it must include at least three value maps: -1 to -1, 0 to 0, and 1 to 1.` 
  The `fromCoordinate` requirements can stay (requiring mappings for -1, 0, and +1), but the `toCoordinate` values could be more free, as long as they don't decrease. A `wght` axis with min/default/max values of 100,400,700 could be mapped to a two-sources system with this avar mapping: `{-1: 0, 0: 0.5, 1: 1}`
